cpp/src/groupby/hash/groupby.cu

```diff
@@ -68,8 +81,248 @@ bool constexpr is_hash_aggregation(aggregation::Kind t) {
+{
+  // flatten the aggs to a table that can be operated on by aggregate_row
+  table_view flattened_values;
+  std::vector<aggregation::Kind> aggs;
+  std::vector<size_t> col_ids;
+  std::tie(flattened_values, aggs, col_ids) = flatten_single_pass_aggs(requests);
+
+  // make table that will hold sparse results
+  std::vector<std::unique_ptr<column>> sparse_columns;
+  for (size_t i = 0; i < aggs.size(); i++) {
+    auto const& col = flattened_values.column(i);
+    bool nullable = (aggs[i] == aggregation::COUNT) ? false : col.has_nulls();
+    auto mask_state = (nullable) ? ALL_NULL : UNALLOCATED;
+    
+    sparse_columns.emplace_back(make_fixed_width_column(
+      experimental::detail::target_type(col.type(), aggs[i]),
+      col.size(), mask_state, stream));
+  }
+  table sparse_table(std::move(sparse_columns));
+  mutable_table_view table_view = sparse_table.mutable_view();
+  experimental::detail::initialize_with_identity(table_view, aggs, stream);
+
+  // prepare to launch kernel to do the actual aggregation
+  auto d_sparse_table = mutable_table_device_view::create(sparse_table);
+  auto d_values = table_device_view::create(flattened_values);
+  rmm::device_vector<aggregation::Kind> d_aggs(aggs);
+
+  bool skip_key_rows_with_nulls = keys_have_nulls and ignore_null_keys;
+
+  experimental::detail::grid_1d grid(keys.num_rows(), 256);
+  if (skip_key_rows_with_nulls) {
+    auto row_bitmask{bitmask_and(keys, rmm::mr::get_default_resource(), stream)};
+    hash::compute_single_pass_aggs<true>
+      <<<grid.num_blocks, grid.num_threads_per_block, 0, stream>>>(
+        map, keys.num_rows(), *d_values, *d_sparse_table, d_aggs.data().get(),
+        static_cast<bitmask_type*>(row_bitmask.data()));
+  } else {
+    hash::compute_single_pass_aggs<false>
+      <<<grid.num_blocks, grid.num_threads_per_block, 0, stream>>>(
+        map, keys.num_rows(), *d_values, *d_sparse_table, d_aggs.data().get(),
+        nullptr);
+  }
+
+  // Add results back to sparse_results cache
+  auto sparse_result_cols = sparse_table.release();
+  for (size_t i = 0; i < aggs.size(); i++) {
+    sparse_results.add_result(col_ids[i],
+                              std::make_unique<aggregation>(aggs[i]),
+                              std::move(sparse_result_cols[i]));
+  }
+}
+
+template <bool keys_have_nulls>
+auto groupby_null_templated(
+    table_view const& keys, std::vector<aggregation_request> const& requests,
+    experimental::detail::result_cache& cache,
+    bool ignore_null_keys, cudaStream_t stream,
+    rmm::mr::device_memory_resource* mr)
+{
+  auto d_keys = table_device_view::create(keys);
+  auto map = create_hash_map<keys_have_nulls>(*d_keys, ignore_null_keys, stream);
+
+  // Cache of sparse results where the location of aggregate value in each
+  // column is indexed by the hash map
+  experimental::detail::result_cache sparse_results(requests.size());
+
+  // Compute all single pass aggs first
+  compute_single_pass_aggs<keys_have_nulls>(
```

Comment:
My reasoning for doing this was that I want to be able to add more aggregations to hash implementation presuming it'd be faster. That means, sometimes, we might not even have to call `compute_single_pass_aggs`. E.g. `nunique` can be done with two hash maps: one for just `keys` - the existing `map`, and one with `keys + vals_col`. Therefore the construction of the map was taken out of `compute_single_pass_aggs` and it was renamed as such.

Reply:
Wouldn't the hash map be different for use in `compute_single_pass_aggs` vs. `nunique`? I.e., the equality and hasher would be different?

It seems like the hash map being returned by `create_hash_map` is specific to use in `compute_single_pass_aggs`, and therefore would make sense to just move creating it inside `compute_single_pass_aggs`.

However, if I'm mistaken and the map can be reused by other functionality, then I think the approach you've taken here makes sense.

Reply:
The `keys` hash map would be the same no? And if there's an aggregation other than `nunique` and one of the single passable ones, then we want their results to be in the same order because the returned `unique_keys` is common.